File: backend/app/realtime_session.py
"""Realtime transcription credentials and server-side relay support.

The browser never sees the standard API key. For the preferred WebRTC
architecture the backend mints an ephemeral client secret; for the fallback
architecture the browser streams PCM16 audio over our own WebSocket and this
module relays it to the provider's realtime endpoint server-side.

NOTE (honest status): this module cannot be exercised in an environment without
an OPENAI_API_KEY and a microphone; it is covered by unit tests with fake
transports and must be smoke-tested against the real API before a live demo.
See MORNING_REVIEW.md.
"""
from __future__ import annotations

import base64
import json
import logging

import httpx

from .config import Settings

logger = logging.getLogger("hormonerx.realtime")


class RealtimeSessionError(Exception):
    pass


async def mint_client_secret(settings: Settings) -> dict:
    """Mint an ephemeral realtime client secret for browser WebRTC use."""
    if not settings.openai_api_key:
        raise RealtimeSessionError("OPENAI_API_KEY is not configured on the server")
    session_config = {
        "session": {
            "type": "transcription",
            "audio": {
                "input": {
                    "transcription": {
                        "model": settings.transcription_model,
                        "language": settings.transcription_language,
                    },
                    "turn_detection": {"type": "server_vad"},
                }
            },
        }
    }
    async with httpx.AsyncClient(timeout=15.0) as client:
        response = await client.post(
            f"{settings.openai_base_url}/realtime/client_secrets",
            headers={"Authorization": f"Bearer {settings.openai_api_key}"},
            json=session_config,
        )
    if response.status_code >= 400:
        logger.error("client secret mint failed: %s %s", response.status_code, response.text[:500])
        raise RealtimeSessionError(f"provider returned {response.status_code}")
    payload = response.json()
    # Return only what the browser needs; never the server key.
    return {
        "client_secret": payload.get("value") or payload.get("client_secret", {}).get("value"),
        "expires_at": payload.get("expires_at") or payload.get("client_secret", {}).get("expires_at"),
        "model": settings.transcription_model,
    }


class ProviderRelay:
    """Server-side relay: our WebSocket audio frames -> provider realtime WS.

    Transport is injected so tests can drive the relay with fake provider events.
    """

    def __init__(self, settings: Settings, on_partial, on_final):
        self.settings = settings
        self.on_partial = on_partial  # async (text) -> None
        self.on_final = on_final  # async (item_id, text) -> None
        self._provider_ws = None

    async def connect(self, websockets_connect=None):
        import websockets

        connect = websockets_connect or websockets.connect
        if not self.settings.openai_api_key:
            raise RealtimeSessionError("OPENAI_API_KEY is not configured on the server")
        url = self.settings.openai_base_url.replace("https://", "wss://") + "/realtime?intent=transcription"
        self._provider_ws = await connect(
            url,
            additional_headers={"Authorization": f"Bearer {self.settings.openai_api_key}"},
        )
        await self._provider_ws.send(
            json.dumps(
                {
                    "type": "transcription_session.update",
                    "session": {
                        "input_audio_transcription": {
                            "model": self.settings.transcription_model,
                            "language": self.settings.transcription_language,
                        },
                        "turn_detection": {"type": "server_vad"},
                    },
                }
            )
        )

    async def send_audio(self, pcm16: bytes) -> None:
        if self._provider_ws is None:
            raise RealtimeSessionError("relay not connected")
        await self._provider_ws.send(
            json.dumps(
                {"type": "input_audio_buffer.append", "audio": base64.b64encode(pcm16).decode("ascii")}
            )
        )

    async def handle_provider_event(self, raw: str) -> None:
        event = json.loads(raw)
        event_type = event.get("type", "")
        if event_type.endswith("input_audio_transcription.delta"):
            await self.on_partial(event.get("delta", ""))
        elif event_type.endswith("input_audio_transcription.completed"):
            await self.on_final(event.get("item_id"), event.get("transcript", ""))

    async def pump(self) -> None:
        if self._provider_ws is None:
            raise RealtimeSessionError("relay not connected")
        async for raw in self._provider_ws:
            await self.handle_provider_event(raw)

    async def close(self) -> None:
        if self._provider_ws is not None:
            await self._provider_ws.close()
            self._provider_ws = None

File: backend/src/realtimeSession.ts
/**
 * Realtime transcription credentials and server-side relay support.
 *
 * The browser never sees the standard API key. For the preferred WebRTC
 * architecture the backend mints an ephemeral client secret; for the fallback
 * architecture the browser streams PCM16 audio over our own WebSocket and this
 * module relays it to the provider's realtime endpoint server-side.
 *
 * NOTE (honest status): this module cannot be exercised in an environment
 * without an OPENAI_API_KEY and a microphone; it is covered by unit tests with
 * fake transports and must be smoke-tested against the real API before a live
 * demo. See MORNING_REVIEW.md.
 */
import WebSocket from 'ws';

import { Settings } from './config.ts';

export class RealtimeSessionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'RealtimeSessionError';
  }
}

/** Mint an ephemeral realtime client secret for browser WebRTC use. */
export async function mintClientSecret(settings: Settings): Promise<Record<string, unknown>> {
  if (!settings.openai_api_key) {
    throw new RealtimeSessionError('OPENAI_API_KEY is not configured on the server');
  }
  const sessionConfig = {
    session: {
      type: 'transcription',
      audio: {
        input: {
          transcription: {
            model: settings.transcription_model,
            language: settings.transcription_language,
          },
          turn_detection: { type: 'server_vad' },
        },
      },
    },
  };
  const response = await fetch(`${settings.openai_base_url}/realtime/client_secrets`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${settings.openai_api_key}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(sessionConfig),
    signal: AbortSignal.timeout(15_000),
  });
  if (response.status >= 400) {
    const text = await response.text();
    console.error(`client secret mint failed: ${response.status} ${text.slice(0, 500)}`);
    throw new RealtimeSessionError(`provider returned ${response.status}`);
  }
  const payload = (await response.json()) as Record<string, any>;
  // Return only what the browser needs; never the server key.
  return {
    client_secret: payload.value ?? payload.client_secret?.value,
    expires_at: payload.expires_at ?? payload.client_secret?.expires_at,
    model: settings.transcription_model,
  };
}

/** Minimal provider-socket contract so tests can inject a fake transport. */
export interface ProviderSocket {
  send(data: string): void | Promise<void>;
  close(): void | Promise<void>;
  on(event: 'message', listener: (data: unknown) => void): void;
  on(event: 'close' | 'error', listener: (arg?: unknown) => void): void;
}

export type ProviderConnect = (url: string, headers: Record<string, string>) => Promise<ProviderSocket>;

async function defaultConnect(url: string, headers: Record<string, string>): Promise<ProviderSocket> {
  const ws = new WebSocket(url, { headers });
  await new Promise<void>((resolve, reject) => {
    ws.once('open', () => resolve());
    ws.once('error', (err) => reject(err));
  });
  return ws as unknown as ProviderSocket;
}

/**
 * Server-side relay: our WebSocket audio frames -> provider realtime WS.
 *
 * Transport is injected so tests can drive the relay with fake provider events.
 */
export class ProviderRelay {
  private providerWs: ProviderSocket | null = null;

  constructor(
    private settings: Settings,
    private onPartial: ((text: string) => void | Promise<void>) | null,
    private onFinal: ((itemId: string | null, text: string) => void | Promise<void>) | null,
  ) {}

  async connect(connectImpl?: ProviderConnect): Promise<void> {
    if (!this.settings.openai_api_key) {
      throw new RealtimeSessionError('OPENAI_API_KEY is not configured on the server');
    }
    const connect = connectImpl ?? defaultConnect;
    const url = this.settings.openai_base_url.replace('https://', 'wss://') + '/realtime?intent=transcription';
    this.providerWs = await connect(url, { Authorization: `Bearer ${this.settings.openai_api_key}` });
    await this.providerWs.send(
      JSON.stringify({
        type: 'transcription_session.update',
        session: {
          input_audio_transcription: {
            model: this.settings.transcription_model,
            language: this.settings.transcription_language,
          },
          turn_detection: { type: 'server_vad' },
        },
      }),
    );
  }

  async sendAudio(pcm16: Buffer | Uint8Array): Promise<void> {
    if (!this.providerWs) {
      throw new RealtimeSessionError('relay not connected');
    }
    await this.providerWs.send(
      JSON.stringify({
        type: 'input_audio_buffer.append',
        audio: Buffer.from(pcm16).toString('base64'),
      }),
    );
  }

  async handleProviderEvent(raw: string): Promise<void> {
    const event = JSON.parse(raw) as { type?: string; delta?: string; item_id?: string; transcript?: string };
    const eventType = event.type ?? '';
    if (eventType.endsWith('input_audio_transcription.delta')) {
      await this.onPartial?.(event.delta ?? '');
    } else if (eventType.endsWith('input_audio_transcription.completed')) {
      await this.onFinal?.(event.item_id ?? null, event.transcript ?? '');
    }
  }

  /** Consume provider events until the provider socket closes. */
  pump(): Promise<void> {
    if (!this.providerWs) {
      return Promise.reject(new RealtimeSessionError('relay not connected'));
    }
    const ws = this.providerWs;
    return new Promise<void>((resolve) => {
      ws.on('message', (data: unknown) => {
        void this.handleProviderEvent(String(data));
      });
      ws.on('close', () => resolve());
      ws.on('error', () => resolve());
    });
  }

  async close(): Promise<void> {
    if (this.providerWs) {
      await this.providerWs.close();
      this.providerWs = null;
    }
  }
}
